components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/lifecycle/LCManager.java

```diff
@@ -0,0 +1,299 @@
+/*
+ *  Copyright (c) 2022, WSO2 LLC. (http://www.wso2.org) All Rights Reserved.
+ *
+ *  WSO2 Inc. licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except
+ *  in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.carbon.apimgt.impl.lifecycle;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+import org.wso2.carbon.apimgt.api.APIManagementException;
+
+import org.wso2.carbon.apimgt.impl.utils.APIUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Manages the LifeCycle
+ */
+public class LCManager {
+
+    private static final Log log = LogFactory.getLog(LCManager.class);
+    private static final String STATE_ID_PROTOTYPED = "Prototyped";
+    private static final String STATE_ID_PUBLISHED = "Published";
+    private static final String TRANSITION_TARGET_PROTOTYPED = "Prototyped";
+    private static final String TRANSITION_TARGET_PUBLISHED = "Published";
+    private static final String LIFECYCLE_KEY = "LifeCycle";
+    private static final String STATES_KEY = "States";
+    private static final String STATE_KEY = "State";
+    private static final String TRANSITIONS_KEY = "Transitions";
+    private static final String CHECK_ITEMS_KEY = "CheckItems";
+    private static final String API_LIFECYCLE_PATH = "lifecycle/APILifeCycle.json";
+    private static final String UTF_8 = "UTF-8";
+    private static final String EVENT_KEY = "Event";
+    private static final String TARGET_KEY = "Target";
+    private Map<String, String> stateTransitionMap = new HashMap<String, String>();
+    private Map<String, StateInfo> stateInfoMap = new HashMap<String, StateInfo>();
+    private HashMap<String, LifeCycleTransition> stateHashMap = new HashMap<String, LifeCycleTransition>();
+    private String tenantDomain;
+    private static JSONObject defaultLCObj;
+    private LCManager instance;
+
+    static {
+        if (defaultLCObj == null) {
+
+            try {
+                defaultLCObj = getDefaultLCConfigJSON();
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```

Comment:
any reason to handle this in static block

Reply:
This was done to reduce the number of file readings. Due to this Static code block the file reading is only done at the start of the program. If we didn't use this static block, this getDefaultLCConfigJSON() would execute on each state change, causing the file readings to occur on each state change, which happens to be the worst case scenario.